pkg/dataflow/affinity.go

```diff
@@ -0,0 +1,121 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// InjectAffinityByRunAfterOp inject the affinity based on preceding operation
+func InjectAffinityByRunAfterOp(c client.Client, runAfter *datav1alpha1.OperationRef, opNamespace string, currentAffinity *v1.Affinity) (*v1.Affinity, error) {
+	// no previous operation or use default affinity strategy, no need to generate node affinity
+	if runAfter == nil || runAfter.AffinityStrategy.Policy == datav1alpha1.DefaultAffinityStrategy {
+		return currentAffinity, nil
+	}
+	precedingOpNamespace := opNamespace
+	if len(runAfter.Namespace) != 0 {
+		precedingOpNamespace = runAfter.Namespace
+	}
+
+	precedingOpStatus, err := utils.GetPrecedingOperationStatus(c, runAfter, precedingOpNamespace)
+	if err != nil {
+		return currentAffinity, err
+	}
+
+	// require policy
+	if runAfter.AffinityStrategy.Policy == datav1alpha1.RequireAffinityStrategy {
+		return injectRequiredAffinity(runAfter, precedingOpStatus.NodeAffinity, currentAffinity)
+	}
+
+	// prefer policy
+	if runAfter.AffinityStrategy.Policy == datav1alpha1.PreferAffinityStrategy {
+		return injectPreferredAffinity(runAfter, precedingOpStatus.NodeAffinity, currentAffinity)
+	}
+
+	return currentAffinity, fmt.Errorf("unknown policy for affinity strategy: %s", runAfter.AffinityStrategy.Policy)
+}
+
+func injectPreferredAffinity(runAfter *datav1alpha1.OperationRef, prevOpNodeAffinity *v1.NodeAffinity, currentAffinity *v1.Affinity) (*v1.Affinity, error) {
+	var preferTerms []v1.PreferredSchedulingTerm
+	prefer := runAfter.AffinityStrategy.Prefer
+	if prefer == nil {
+		prefer = []datav1alpha1.Prefer{
+			{
+				Name:   common.K8sNodeNameLabelKey,
+				Weight: 100,
+			},
+		}
+	}
+	// currently, only has one element.
+	podNodeSelectorTerm := prevOpNodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms[0]
+
+	for _, item := range prefer {
+		key := item.Name
+		for _, expression := range podNodeSelectorTerm.MatchExpressions {
+			if expression.Key == key {
+				preferTerms = append(preferTerms, v1.PreferredSchedulingTerm{
+					Weight: item.Weight,
+					Preference: v1.NodeSelectorTerm{
+						MatchExpressions: []v1.NodeSelectorRequirement{
+							{
+								Key:      expression.Key,
+								Operator: expression.Operator,
+								Values:   expression.Values,
+							},
+						},
+					},
+				})
+			}
+		}
+	}
+	return utils.InjectPreferredSchedulingTermsToAffinity(preferTerms, currentAffinity), nil
+}
+
+func injectRequiredAffinity(runAfter *datav1alpha1.OperationRef, prevOpNodeAffinity *v1.NodeAffinity, currentAffinity *v1.Affinity) (*v1.Affinity, error) {
+	if prevOpNodeAffinity == nil || prevOpNodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution == nil ||
+		prevOpNodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms == nil {
+		return currentAffinity, nil
+	}
+
+	var matchExpressions []v1.NodeSelectorRequirement
+	keys := runAfter.AffinityStrategy.Require
+	if keys == nil {
```

Comment:
How about using len(keys) == 0 for a more wider case?

Reply:
fixed.